includes/admin/class-wc-rest-payments-reader-controller.php

```diff
@@ -62,4 +100,69 @@ public function get_summary( $request ) {
 		return rest_ensure_response( $summary );
 	}
 
+	/**
+	 * Renders HTML for a print receipt
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_HTTP_RESPONSE|WP_Error
+	 */
+	public function generate_print_receipt( $request ) {
+		try {
+			$payment_intent = $this->api_client->get_intent( $request->get_param( 'payment_id' ) );
+
+			if ( 'succeeded' !== $payment_intent->get_status() ) {
+				return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', __( 'Invalid payment', 'woocommerce-payments' ), [ 'status' => 500 ] ) );
+			}
+
+			$charge = $this->api_client->get_charge( $payment_intent->get_charge_id() );
+		} catch ( API_Exception $e ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', $e->getMessage(), [ 'status' => $e->get_http_code() ] ) );
+		}
+
+		$order = wc_get_order( $charge['order']['number'] );
+
+		if ( ! $order ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', __( 'Order not found', 'woocommerce-payments' ), [ 'status' => 500 ] ) );
+		}
+
+		try {
+			$settings = [
+				'business_name' => $this->wcpay_gateway->get_option( 'account_business_name' ),
+				'support_info'  => [
+					'address' => $this->wcpay_gateway->get_option( 'account_business_support_address' ),
+					'phone'   => $this->wcpay_gateway->get_option( 'account_business_support_phone' ),
+					'email'   => $this->wcpay_gateway->get_option( 'account_business_support_email' ),
+				],
+			];
+		} catch ( Exception $e ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', __( 'There was a problem retrieving merchant settings.', 'woocommerce-payments' ), [ 'status' => 500 ] ) );
+		}
+
+		try {
+			$receipt_data = $this->receipts_service->get_receipt_markup( $settings, $order, $charge );
+		} catch ( \Throwable $th ) {
+			return rest_ensure_response( new WP_Error( 'generate_print_receipt_error', $th->getMessage(), [ 'status' => 500 ] ) );
+		}
+
+		/**
+		 * WP_REST_Server will convert the response data to JSON prior to output it.
+		 * Using this filter to prevent it, and output the data from WP_HTTP_Response instead.
+		 */
+		add_filter(
```

Comment:
@mgascam, why do we need this filter (I'm not really getting it)?

Reply:
@kalessil  This is related to the fact that I need the response of this endpoint to be HTML instead of JSON. As far as I understand, the default behavior is to convert to JSON the data of the response. But in this case I want the response to be HTML.

 If i follow the flow correctly, this happens in the `serve_request` method of `WP_REST_Server` [link](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/class-wp-rest-server.php#L495). 

Since I am not interested in that behavior I am using the `rest_pre_serve_request`, that allows me to `echo` the response without converting it to JSON.

I guess we could go for a different approach and not use this filter if we return the string representation of the html inside the JSON payload. Something like:

`
{
    'receipt': '<html><body>The receipt</body></html>
}
`

I hope I explained myself :)

Reply:
Yes, thanks!